README.md
## <Pre>serve

Preserves content inside `<pre>`, `<code>`, `<samp>` tags.

- Also compatible with Jetpack Markdown; i.e., fenced code blocks and inline `code`.

`<Pre>`serve Pro is a commercial plugin available at [https://wpsharks.com/product/pre-serve-pro](wpsharks.com). The plugin is hosted here on a public Github repository in order to better facilitate community contributions from developers and users alike. If you have a suggestion, a bug report, or a patch for an issue, feel free to submit it here. We do ask, however, that if you are using the plugin on a live site that you please purchase a valid license from the [https://wpsharks.com/product/pre-serve-pro](wpsharks.com). We cannot provide support to anyone that does not hold a valid license key.


CHANGELOG.md
## v160919.19112

- Preserving `get_the_excerpt()` also.

## v160731.37913

- Updating to the latest WP Sharks core.

## v160722.57918

- Initial release.


plugin.php
<?php
/**
 * Plugin.
 *
 * @wp-plugin
 *
 * Version: 160919.19112
 * Text Domain: pre-serve
 * Plugin Name: &lt;Pre&gt;serve Pro
 *
 * Author: WP Sharks™
 * Author URI: https://wpsharks.com
 *
 * License: GPL-3.0+
 * License URI: http://www.gnu.org/licenses/gpl-3.0.html
 *
 * Plugin URI: https://wpsharks.com/product/pre-serve-pro
 * Description: Preserves content inside &lt;pre&gt;, &lt;code&gt;, or &lt;samp&gt; tags.
 */
// PHP v5.2 compatible.

if (!defined('WPINC')) {
    exit('Do NOT access this file directly: '.basename(__FILE__));
}
require dirname(__FILE__).'/src/includes/wp-php-rv.php';

if (require(dirname(__FILE__).'/src/vendor/websharks/wp-php-rv/src/includes/check.php')) {
    require_once dirname(__FILE__).'/src/includes/plugin.php';
} else {
    wp_php_rv_notice('<Pre>serve Pro');
}


src/includes/classes/Base/Facades.php
<?php
/**
 * Facades.
 *
 * @author @jaswsinc
 * @copyright WP Sharks™
 */
declare (strict_types = 1);
namespace WebSharks\WpSharks\Preserve\Classes\Base;

use WebSharks\WpSharks\Preserve\Classes;
use WebSharks\WpSharks\Preserve\Interfaces;
use WebSharks\WpSharks\Preserve\Traits;
#
use WebSharks\WpSharks\Preserve\Classes\AppFacades as a;
use WebSharks\WpSharks\Preserve\Classes\SCoreFacades as s;
use WebSharks\WpSharks\Preserve\Classes\CoreFacades as c;
#
use WebSharks\WpSharks\Core\Classes as SCoreClasses;
use WebSharks\WpSharks\Core\Interfaces as SCoreInterfaces;
use WebSharks\WpSharks\Core\Traits as SCoreTraits;
#
use WebSharks\Core\WpSharksCore\Classes as CoreClasses;
use WebSharks\Core\WpSharksCore\Classes\Core\Base\Exception;
use WebSharks\Core\WpSharksCore\Interfaces as CoreInterfaces;
use WebSharks\Core\WpSharksCore\Traits as CoreTraits;
#
use function assert as debug;
use function get_defined_vars as vars;

/**
 * Pseudo-static facades.
 *
 * @since 160722.57589 Initial release.
 */
abstract class Facades
{
}


src/includes/classes/Utils/Content.php
<?php
/**
 * Content utils.
 *
 * @author @jaswsinc
 * @copyright WP Sharks™
 */
declare (strict_types = 1);
namespace WebSharks\WpSharks\Preserve\Classes\Utils;

use WebSharks\WpSharks\Preserve\Classes;
use WebSharks\WpSharks\Preserve\Interfaces;
use WebSharks\WpSharks\Preserve\Traits;
#
use WebSharks\WpSharks\Preserve\Classes\AppFacades as a;
use WebSharks\WpSharks\Preserve\Classes\SCoreFacades as s;
use WebSharks\WpSharks\Preserve\Classes\CoreFacades as c;
#
use WebSharks\WpSharks\Core\Classes as SCoreClasses;
use WebSharks\WpSharks\Core\Interfaces as SCoreInterfaces;
use WebSharks\WpSharks\Core\Traits as SCoreTraits;
#
use WebSharks\Core\WpSharksCore\Classes as CoreClasses;
use WebSharks\Core\WpSharksCore\Classes\Core\Base\Exception;
use WebSharks\Core\WpSharksCore\Interfaces as CoreInterfaces;
use WebSharks\Core\WpSharksCore\Traits as CoreTraits;
#
use function assert as debug;
use function get_defined_vars as vars;

/**
 * Content utils.
 *
 * @since 160722.57589 Content utils.
 */
class Content extends SCoreClasses\SCore\Base\Core
{
    /**
     * Tokenizers.
     *
     * @since 160722.57589 Content utils.
     *
     * @var array|null
     */
    protected $Tokenizers;

    /**
     * Preserve pre/code/samp.
     *
     * @since 160722.57589 Content utils.
     *
     * @param string|scalar $content Content.
     *
     * @return string $content Filtered content.
     */
    public function onTheContentPreserve($content): string
    {
        $content = (string) $content;

        if (!$content) {
            $this->Tokenizers[] = null;
            return $content; // Nothing to do.
        } elseif (mb_strpos($content, '[') === false) {
            $this->Tokenizers[] = null;
            return $content; // Nothing to do.
        } elseif (!preg_match('/\<(?:pre|code|samp)/ui', $content)) {
            $this->Tokenizers[] = null;
            return $content; // Nothing to do.
        }
        $Tokenizer          = c::tokenize($content, ['pre', 'code', 'samp']);
        $this->Tokenizers[] = $Tokenizer; // End of stack.
        return $content     = $Tokenizer->getString();
    }

    /**
     * Restore pre/code/samp.
     *
     * @since 160722.57589 Content utils.
     *
     * @param string|scalar $content Content.
     *
     * @return string $content Filtered content.
     */
    public function onTheContentRestore($content): string
    {
        $content = (string) $content;

        if (!$this->Tokenizers) {
            debug(0, c::issue(vars(), 'Missing tokenizers.'));
            return $content; // Nothing to do.
        } elseif (!($Tokenizer = array_pop($this->Tokenizers))) {
            return $content; // Nothing to do.
        } // Also pops last tokenizer off the stack.

        $Tokenizer->setString($content); // As it exists now.

        return $content = $Tokenizer->restoreGetString();
    }
}


src/includes/classes/App.php
<?php
/**
 * Application.
 *
 * @author @jaswsinc
 * @copyright WP Sharks™
 */
declare (strict_types = 1);
namespace WebSharks\WpSharks\Preserve\Classes;

use WebSharks\WpSharks\Preserve\Classes;
use WebSharks\WpSharks\Preserve\Interfaces;
use WebSharks\WpSharks\Preserve\Traits;
#
use WebSharks\WpSharks\Preserve\Classes\AppFacades as a;
use WebSharks\WpSharks\Preserve\Classes\SCoreFacades as s;
use WebSharks\WpSharks\Preserve\Classes\CoreFacades as c;
#
use WebSharks\WpSharks\Core\Classes as SCoreClasses;
use WebSharks\WpSharks\Core\Interfaces as SCoreInterfaces;
use WebSharks\WpSharks\Core\Traits as SCoreTraits;
#
use WebSharks\Core\WpSharksCore\Classes as CoreClasses;
use WebSharks\Core\WpSharksCore\Classes\Core\Base\Exception;
use WebSharks\Core\WpSharksCore\Interfaces as CoreInterfaces;
use WebSharks\Core\WpSharksCore\Traits as CoreTraits;
#
use function assert as debug;
use function get_defined_vars as vars;

/**
 * Application.
 *
 * @since 160722.57589 Initial release.
 */
class App extends SCoreClasses\App
{
    /**
     * Version.
     *
     * @since 160722.57589 Initial release.
     *
     * @var string Version.
     */
    const VERSION = '160919.19112'; //v//

    /**
     * Constructor.
     *
     * @since 160722.57589 Initial release.
     *
     * @param array $instance Instance args.
     */
    public function __construct(array $instance = [])
    {
        $instance_base = [
            '©di' => [
                '©default_rule' => [
                    'new_instances' => [
                    ],
                ],
            ],

            '§specs' => [
                '§type' => 'plugin',
                '§file' => dirname(__FILE__, 4).'/plugin.php',
            ],
            '©brand' => [
                '©acronym' => 'PREServe',
                '©name'    => '<Pre>serve',

                '©slug' => 'pre-serve',
                '©var'  => 'pre_serve',

                '©short_slug' => 'pre-serve',
                '©short_var'  => 'pre_serve',

                '©text_domain' => 'pre-serve',
            ],

            '§pro_option_keys' => [],
            '§default_options' => [],
        ];
        parent::__construct($instance_base, $instance);
    }

    /**
     * Other hook setup handler.
     *
     * @since 160722.57589 Initial release.
     */
    protected function onSetupOtherHooks()
    {
        parent::onSetupOtherHooks();

        # On `init` w/ a late priority.

        add_action('init', function () {

            # Filters `the_content`. Covers most WP themes/plugins.

            add_filter('the_content', [$this->Utils->Content, 'onTheContentPreserve'], -1000);
            add_filter('the_content', [$this->Utils->Content, 'onTheContentRestore'], 1000);

            # Filters `get_the_excerpt`. For WP themes/plugins that use a true excerpt.

            add_filter('get_the_excerpt', [$this->Utils->Content, 'onTheContentPreserve'], -1000);
            add_filter('get_the_excerpt', [$this->Utils->Content, 'onTheContentRestore'], 1000);

            # Filters `if_shortcode_content` for compatibility w/ the `[if]` shortcode.

            add_filter('if_shortcode_content', [$this->Utils->Content, 'onTheContentPreserve'], -1000);
            add_filter('if_shortcode_content', [$this->Utils->Content, 'onTheContentRestore'], 1000);

            # Filters `woocommerce_short_description` for compatibility w/ WooCommerce.

            if (has_filter('woocommerce_short_description', 'wc_format_product_short_description') && s::jetpackCanMarkdown()) {
                // Fixing a WooCommerce bug by Moving Jetpack markdown to a more logical/compatible priority.
                // I consider this a bug because WooCommerce applies markdown 'after' other filters, which can cause corruption.
                // WooCommerce also applies `wpautop()` twice. Once via filter, then again after a late/buggy markdown.
                if (remove_filter('woocommerce_short_description', 'wc_format_product_short_description', 9999999)) {
                    add_filter('woocommerce_short_description', s::class.'::jetpackMarkdown', -10000);
                } else { // Flag this as a potential problem whenver debugging is enabled by a developer.
                    debug(0, c::issue([], 'Failed to remove `wc_format_product_short_description()` filter.'));
                }
            } // Preservation comes after Jetpack markdown. Markdown should almost always occur first!
            add_filter('woocommerce_short_description', [$this->Utils->Content, 'onTheContentPreserve'], -1000);
            add_filter('woocommerce_short_description', [$this->Utils->Content, 'onTheContentRestore'], 1000);
        }, 100);
    }
}
